### ISSUE TITLE
React-Calendar: Throttled when calendar has many entries

### ISSUE THREAD

**Issue by AlanSharePoint** [pnp/sp-dev-fx-webparts#1833]
React-Calendar: Throttled when calendar has many entries

Hi Team,
@mohammadamer , @Abderahman88 , 

## Expected or Desired Behavior
We are using the React-Calendar heavily on our main intranet page. When the page is reloaded multiple times, such as when updating and editing the page, authors experience a throttle notification from SharePoint and are unable to use SharePoint until their throttle times out. If we remove the React-Calendar web part, users no longer get throttled. Throttling is unexpected behavior. 

## Observed Behavior
User is throttled when page is refreshed multiple times with a heavily-used calendar. Removing the react-calendar remedied the throttle issue. 

## Steps to Reproduce
We have 85 entries in the month view of the calendar. The calendar web part is displayed in month view on a SharePoint o365 page. 
Refresh page multiple times in a row, or edit/save page multiple times in a row. User is then throttled by SharePoint Online. 

# Environment Details (*Development & Target environment*)

- **OS**: Windows 10
- **Target Environment**: SharePoint Online
- **Browser(s)**: All
- **Additional details**: The calendar works great when there are less entries, but once it gets multiple entries, it is slow to load and then we receive throttle limits. 

## Additional context
Please reach out if you require additional information. Thanks so much!

**Comment by Abderahman88** [comment]
Thanks @AlanSharePoint to report this issue. I'll check it out 👍

**Comment by Abderahman88** [comment]
Hi @AlanSharePoint,

I think it's slow, because of all the timezone conversions.

` await this.getLocalTime(event.EventDate)`

In Chrome network-tab, I see a lot of `utctolocaltime` requests

![image](https://user-images.githubusercontent.com/36161889/116910927-90aee000-ac46-11eb-9eda-0e3bc5e2cd95.png)

I'm not sure how to solve this issue at the moment. Anyone an idea?

**Comment by Eli-Schei** [comment]
I had a look at this issue, and here are som things I found.

### **TL;DR;** 
Updated to newest version of pnpjs to use it to get the local timezone once.  But this update require a lot of refactoring of the code. So a much larger job. 
Rolled back to orignal and added a function that saves the events to localstorage and only calls the api if there are any new events on refresh - but this made the "week-view" break 


### The long version:
For every event in the calendar, that the current user have permission to see, the  ` await sp.web.regionalSettings.timeZone.utcToLocalTime ` call is made twice - once for the start-date, and once for the end-date. And this happens on every refresh.  So if there are a lot of events there are also A LOT of calls. 

I thought I would be able to fix this by getting the localTimeZone once - and using it in the loop instead of doing the api-call for every event in the loop. But the `const s = await sp.web.regionalSettings.timeZone();` (from documentation https://pnp.github.io/pnpjs/sp/regional-settings/) didn't work. I figured this was probably due to an older pnpjs version, so I updated to a newer one. But now a lot of the imports are failing so it requires a larger refactoring of the entire webpart.

Instead of refactoring the whole webpart I rolled back to the original version and added a function that saves the events in localstorage. On page-refresh it will compare the local storage version with the latest lists of events. If they are equal the webpart will use the local-storage version to display the events - so no utcToLocalTime calls needed.  And if the are not equal the events will be mapped (and the utcToLocalTime called for every event). 

This works well, exept that the week-view suddenly stopped working. 

Since I'm quite busy for the next couple of weeks I thought I would just leave a comment here with my findings in case someone else wants to have a go. 

You can have a look at my fork to see what I changed: https://github.com/Eli-Schei/sp-dev-fx-webparts/commit/76c8cbcd96a90a94a7a6bda60752f790a9ee6712 

Also here are some screenshots of the error that occurs when week-view is selected. The other views still work fine. 
In gui this is the error message: 
<img width="427" alt="forwardRef" src="https://user-images.githubusercontent.com/60283283/117648699-d7dd2980-b18e-11eb-83fd-ff9628b0d481.png">

And in console this is the error:
<img width="329" alt="n-get" src="https://user-images.githubusercontent.com/60283283/117648698-d7449300-b18e-11eb-9995-330c60246987.png">
And if you click the error in console this is the function that shows up (in dist/calendar-wep-part.js). 
<img width="266" alt="createAccessor" src="https://user-images.githubusercontent.com/60283283/117648700-d875c000-b18e-11eb-980f-e4222dc00f53.png">


**Note**: Should consider to add propper throttle handling too, not just 'work around'  (if 429 retry after... and so on)

**Comment by Eli-Schei** [comment]
Typically like 2minutes after making the above comment I figured out what was causing the week-view error. I just had to move two lines of code. 
Updated commit can be found here--> https://github.com/Eli-Schei/sp-dev-fx-webparts/commit/894ed03ca08b9e05075fcaa987742d386b378b66 .
So now the local storage solutions works fine which will reduce the number of times the api is called - and therefore reduce the risk of being throttled because of frequent page-refresh. But someone should probably add throttle-handling also . :) 
Should I create a pull-request for the fix I have made for now @hugoabernier ?

**Comment by hugoabernier** [comment]
@Eli-Schei yes please!

**Comment by mohammadamer** [comment]
Awesome, Great job @Eli-Schei

**Comment by hugoabernier** [comment]
@AlanSharePoint the latest PR from @Eli-Schei should resolve the issue. Thanks @Eli-Schei for your help with this!

**Comment by AlanSharePoint** [comment]
@hugoabernier @Eli-Schei , thank you so much! We have installed the update, and it loads so much faster. Load time was around 60 seconds and is now down to about 3 seconds. Thanks again!

**Comment by Eli-Schei** [comment]
@AlanSharePoint So glad it helped! :)

**Comment by mohammadamer** [comment]
Hi @Eli-Schei

I just noticed that after pull this change to my calendar WebPart that first load of events is working perfect but second load which is getting events from local storage made a change to events dates.
EX: If the start and end dates of the event are (All day event)
Fri July 09 2021 00:00:00
Fri July 09 2021 23:59:00

It will show up in the calendar view like
Fri July 09 2021 21:00:00
Sat July 10 2021 20:59:00

Why, because the event dates in the calendarEventsWithLocalTime local storage are
EventDate: "2021-07-09T21:00:00.000Z"
EndDate: "2021-07-10T20:59:00.000Z"

Could you please check that issue.
